packages/kokkos/core/src/impl/Kokkos_Core.cpp

```diff
@@ -61,8 +61,64 @@ std::stack<std::function<void()> > finalize_hooks;
 
 namespace Kokkos { namespace Impl { namespace {
 
-int get_ctest_gpu(const char* ctest_process_count_str)
+int get_ctest_gpu(const char* local_rank_str, const char* ctest_kokkos_device_type, const char* ctest_process_count_str)
 {
+  // Make sure rank is within bounds of processes specified by CTest
+  auto process_count = std::atoi(ctest_process_count_str);
+  auto local_rank = std::atoi(local_rank_str);
+  if (local_rank >= process_count) {
+    return 0;
+  }
+
+  // Get the resource types allocated to this process
+  std::ostringstream ctest_process;
+  ctest_process << "CTEST_PROCESS_" << local_rank;
+  auto ctest_process_str = std::getenv(ctest_process.str().c_str());
+  if (!ctest_process_str) {
+    return 0;
+  }
+
+  // Look for the device type specified in CTEST_KOKKOS_DEVICE_TYPE
+  bool found_device = false;
+  auto begin = ctest_process_str;
+  auto comma = std::strchr(begin, ',');
```

Comment:
If you're writing C-style code, then you could just use `strtok`.

Reply:
I'd very much prefer not to use the original `strtok`. It is neither thread-safe nor reentrant, and using it is generally bad practice. I am open to using `strtok_s`, though it requires C11 and seems to only be available through `<string.h>` (not `<cstring>`).

Reply:
If you want a C++ replacement for strtok, especially since there's only comma for delimiter, you could do something like this. There's probably a tighter way to check for the end of string but this is how I always do it.
```
std::istringstream is(begin);
while(true)
{
    std::string devName;
    std::getline(is, devName, ',');
    if(is.eof() || devName.length() == 0)
    {
        break;
    }
    if(devName == ctest_kokkos_device_type)
    {
        found_device = true;
        break;
    }
}
```

Reply:
I like this idea (though abusing `std::getline` seems a little hacky to me.)

Reply:
I don't particularly like `strtok` -- it just seems like the code would benefit from a "tokenize this string" abstraction.

Reply:
Agreed, and I ended up going with @brian-kelley's `std::getline()` idea.